src/System.Management.Automation/engine/SessionStateLocationAPIs.cs

```diff
@@ -590,6 +590,79 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context, bool l
             return this.CurrentLocation;
         }
 
+        /// <summary>
+        /// Changes the current working directory to the path specified.
+        /// It is only for FileSystem provider paths and optimized to be fast.
+        /// </summary>
+        /// <param name="path">
+        /// The file system path of the new current working directory.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        /// If <paramref name="path"/> is null.
+        /// </exception>
+        /// <exception cref="DriveNotFoundException">
+        /// If <paramref name="path"/> refers to a drive that does not exist.
+        /// </exception>
+        internal void SetLocationFileSystemFast(string path)
+        {
+            if (path == null)
+            {
+                throw PSTraceSource.NewArgumentNullException(nameof(path));
+            }
+
+            PathInfo current = null;
+            if (PublicSessionState.InvokeCommand.LocationChangedAction != null)
+            {
+                // It is under a condition to avoid the unneeded allocation - CurrentLocation always allocates.
+                current = CurrentLocation;
+            }
+
+            PSDriveInfo previousWorkingDrive = CurrentDrive;
+
+            if (LocationGlobber.IsProviderDirectPath(path))
+            {
+                // The path is a provider-direct (drive is implicitly in the provider path) path so use the current
+                // provider and its hidden drive but don't modify the path
+                // at all.
+                ProviderInfo provider = CurrentLocation.Provider;
+                CurrentDrive = provider.HiddenDrive;
+            }
+            else
+            {
+                // See if the path is a relative or absolute
+                // path.
+                if (Globber.IsAbsolutePath(path, out string driveName))
+                {
+                    // Since the path is an absolute path
+                    // we need to change the current working
+                    // drive
+                    CurrentDrive = GetDrive(driveName);
+
+                    // Now that the current working drive is set,
+                    // process the rest of the path as a relative path.
+                }
+            }
+
+            string currentPath = path.Substring(CurrentDrive.Root.Length);
```

Comment:
If current drive is Temp:/ we always call IsAbsolutePath() and always set current drive based on the `path`. I checked this for shorter code path that is ResetRunspaceState(). Other code paths call previously Bind(), there is a call SetSessionStateDrive(), then again SetSessionStateDrive() and we fall in IsAbsolutePath() too.

Reply:
See the following repro:
```powershell
cd temp:\
mkdir test
$t =$ExecutionContext.SessionState.GetType()
$p = $t.GetProperties("nonpublic, instance")
$p = $p[0]
$v = $p.GetValue($ExecutionContext.SessionState)
$t = $v.GetType()

## Get the existing 'SetLocation(string)' method
$m1 = $t.GetMethod("SetLocation", [System.Reflection.BindingFlags]"nonpublic, instance", $null, [type[]]@([string]), $null)
## Success. Location is set to 'Temp:\test'
$m1.Invoke($v, @("Temp:\test"))

## Get the new 'SetLocationFileSystemFast' method and invoke it.
$m2 = $t.GetMethod("SetLocationFileSystemFast", [System.Reflection.BindingFlags]"nonpublic, instance")
## Fail with 'ArgumentOutOfRangeException'
$m2.Invoke($v, @("Temp:\test"))
> Exception calling "Invoke" with "2" argument(s): "startIndex cannot be larger than length of string. (Parameter 'startIndex')"
```
I know you have examined the current existing call sites of this method and know they are safe, but the fast-code-path method is internal, and it could be used by someone in other places in future, without knowing the prerequisites or assumptions for this method to work. This is not the right pattern to follow, it will results code hard to maintain.